Sources/GRPC/ConnectionManager.swift

```diff
@@ -0,0 +1,586 @@
+
+      case .transientFailure(let state):
+        return state.readyChannelPromise.futureResult
+
+      case .shutdown:
+        return self.eventLoop.makeFailedFuture(GRPCStatus(code: .unavailable, message: nil))
+      }
+    }
+  }
+
+  /// Shutdown any connection which exists. This is a request from the application.
+  internal func shutdown() -> EventLoopFuture<Void> {
+    return self.eventLoop.flatSubmit {
+      let shutdown: ShutdownState
+
+      switch self.state {
+      // We don't have a channel and we don't want one, easy!
+      case .idle:
+        shutdown = ShutdownState(closeFuture: self.eventLoop.makeSucceededFuture(()))
+        self.state = .shutdown(shutdown)
+
+      // We're mid-connection: the application doesn't have any 'ready' channels so we'll succeed
+      // the shutdown future and deal with any fallout from the connecting channel without the
+      // application knowing.
+      case .connecting(let state):
+        shutdown = ShutdownState(closeFuture: self.eventLoop.makeSucceededFuture(()))
+        self.state = .shutdown(shutdown)
+
+        // Fail the ready channel promise: we're shutting down so even if we manage to successfully
+        // connect the application shouldn't should have access to the channel.
+        state.readyChannelPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+        // In case we do successfully connect, close immediately.
+        state.candidate.whenSuccess {
+          $0.close(mode: .all, promise: nil)
+        }
+
+      // We have an active channel but the application doesn't know about it yet. We'll do the same
+      // as for `.connecting`.
+      case .active(let state):
+        shutdown = ShutdownState(closeFuture: self.eventLoop.makeSucceededFuture(()))
+        self.state = .shutdown(shutdown)
+
+        // Fail the ready channel promise: we're shutting down so even if we manage to successfully
+        // connect the application shouldn't should have access to the channel.
+        state.readyChannelPromise.fail(GRPCStatus(code: .unavailable, message: nil))
+        // We have a channel, close it.
+        state.candidate.close(mode: .all, promise: nil)
+
+      // The channel is up and running: the application could be using it. We can close it and
+      // return the `closeFuture`.
+      case .ready(let state):
+        shutdown = ShutdownState(closeFuture: state.channel.closeFuture)
+        self.state = .shutdown(shutdown)
+
+        // We have a channel, close it.
+        state.channel.close(mode: .all, promise: nil)
+
+      // Like `.connecting` and `.active` the application does not have a `.ready` channel. We'll
+      // do the same but also cancel any scheduled connection attempts and deal with any fallout
+      // if we cancelled too late.
+      case .transientFailure(let state):
+        // Stop the creation of a new channel, if we can.
+        state.scheduled.cancel()
```

Comment:
Given that this `cancel` may fail, do we need to handle the possibility the connection attempt goes ahead anyway?

Reply:
Good question: I think the comment may actually be wrong here; I don't think the cancel can fail: the scheduled task is `startConnecting()`: if that gets run the actual connection attempt will start _after_ the transition from `transientFailure` to `.connecting` (because we submit the connect onto the EventLoop).

Reply:
I mean, `cancel` can always fail. The question is whether we tolerate its failure. I think the answer now is yes, but we should say what we expect to happen here.